saleor/checkout/utils.py

```diff
@@ -57,16 +649,162 @@ def _get_product_or_category_voucher_discount_for_checkout(voucher, checkout):
 
     Raise NotApplicable if voucher of given type cannot be applied.
     """
     if voucher.type == VoucherType.VALUE:
-        return _get_value_voucher_discount_for_checkout(voucher, checkout)
+        return _get_value_voucher_discount_for_cart(voucher, cart)
     if voucher.type == VoucherType.SHIPPING:
-        return _get_shipping_voucher_discount_for_checkout(voucher, checkout)
+        return _get_shipping_voucher_discount_for_cart(voucher, cart)
     if voucher.type in (VoucherType.PRODUCT, VoucherType.CATEGORY):
-        return _get_product_or_category_voucher_discount_for_checkout(
-            voucher, checkout)
+        return _get_product_or_category_voucher_discount_for_cart(
+            voucher, cart)
     raise NotImplementedError('Unknown discount type')
+
+
+def get_voucher_for_cart(cart, vouchers=None):
+    """Return voucher with voucher code saved in cart if active or None."""
+    if cart.voucher_code is not None:
+        if vouchers is None:
+            vouchers = Voucher.objects.active(date=date.today())
+        try:
+            return vouchers.get(code=cart.voucher_code)
+        except Voucher.DoesNotExist:
+            return None
+    return None
+
+
+def recalculate_cart_discount(cart):
+    """Recalculate `cart.discount` based on the voucher.
+
+    Will clear both voucher and discount if the discount is no longer
+    applicable.
+    """
+    voucher = get_voucher_for_cart(cart)
+    if voucher is not None:
+        try:
+            cart.discount_amount = get_voucher_discount_for_cart(voucher, cart)
+            cart.discount_name = voucher.name
+            cart.save(update_fields=['discount_amount', 'discount_name'])
+        except NotApplicable:
+            remove_voucher_from_cart(cart)
+    else:
+        remove_voucher_from_cart(cart)
+
+
+def remove_voucher_from_cart(cart):
+    """Remove voucher data from cart."""
+    cart.voucher_code = None
+    cart.discount_name = None
+    cart.discount_amount = ZERO_MONEY
+    cart.save(update_fields=[
+        'voucher_code', 'discount_name', 'discount_amount'])
+
+
+def get_taxes_for_cart(cart, default_taxes):
+    """Return taxes (if handled) due to shipping address or default one."""
+    if not settings.VATLAYER_ACCESS_KEY:
+        return None
+
+    if cart.shipping_address:
+        return get_taxes_for_country(cart.shipping_address.country)
+
+    return default_taxes
+
+
+def check_shipping_method(cart):
+    """Check if shipping method is valid for address and remove (if not)."""
+    country_code = cart.shipping_address.country.code
+    valid_methods = ShippingMethodCountry.objects.select_related(
+        'shipping_method').unique_for_country_code(country_code)
+    if cart.shipping_method not in valid_methods:
+        cart.shipping_method = None
+        cart.save()
+        return False
+    return True
+
+
+@transaction.atomic
+def create_order(cart, tracking_code, discounts, taxes):
```

Comment:
This function is almost 80 lines long, can we introduce some type of helpers here?

Reply:
You are right, it was hard to read. I added some helpers.